Remove Slow Downloads stuck on 0kb size / MetaData Not being detected as slow.

slkiny opened the pull request:
<img width="1241" alt="image" src="https://github.com/ManiMatter/decluttarr/assets/128746884/56b2e899-4c2f-40bb-a352-d85d9b2becbc">

Currently these use cases fail to be detected. For torrents which are not very popular you may fail to even download the magnet file.

<img width="1241" alt="image" src="https://github.com/ManiMatter/decluttarr/assets/128746884/0ccc9e9c-a263-4efe-94d3-6263c6e62dfc">

Instead it was detected as a big download that might be moving soon... This will fix the issue, and ensure that something is downloaded.


Example JSON of a Magnet File that is not downloaded.

`[INFO]:  {'seriesId': 90, 'episodeId': 8097, 'seasonNumber': 4, 'languages': [], 'quality': {'quality': {'id': 5, 'name': 'WEBDL-720p', 'source': 'web', 'resolution': 720}, 'revision': {'version': 1, 'real': 0, 'isRepack': False}}, 'customFormats': [], 'customFormatScore': 0, 'size': 0, 'title': 'fc2e9e29471ed570771be8dfa157c512fad6e6eb', 'sizeleft': 0, 'added': '2024-06-10T00:16:14Z', 'status': 'downloading', 'trackedDownloadStatus': 'ok', 'trackedDownloadState': 'downloading', 'statusMessages': [], 'downloadId': 'FC2E9E29471ED570771BE8DFA157C512FAD6E6EB', 'protocol': 'torrent', 'downloadClient': 'qbit', 'downloadClientHasPostImportCategory': False, 'indexer': 'TorrentDownload (Prowlarr)', 'episodeHasFile': False, 'id': 333611039} 
`


<img width="1241" alt="image" src="https://github.com/ManiMatter/decluttarr/assets/128746884/a4e47f92-603e-46ba-a373-93c88fc93c9f">


Example logs when it is working.

ManiMatter commented:
Hi @slkiny 

welcome and thank you for your PR.

I am not sure I understand what is happening here. On the picture below, the downloads are in "Downloading metatada" stage. On the json that you posted, they are in  `'status': 'downloading', 'trackedDownloadStatus': 'ok', 'trackedDownloadState': 'downloading', 'statusMessages': []`. So no mentioning of metdata.
> <img alt="image" width="1241" src="https://private-user-images.githubusercontent.com/128746884/338023187-56b2e899-4c2f-40bb-a352-d85d9b2becbc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgwMDE3NjEsIm5iZiI6MTcxODAwMTQ2MSwicGF0aCI6Ii8xMjg3NDY4ODQvMzM4MDIzMTg3LTU2YjJlODk5LTRjMmYtNDBiYi1hMzUyLWQ4NWQ5YjJiZWNiYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxMFQwNjM3NDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05Yzk5M2FkZDYwMzc0NDM1YjAxOWY4ZGZlNDA0OWE5ZGI1ZmUyMGM2NzFhZDViNTU3YTkyMThmZTBlNmE1Yjg3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.8bIyyZXlSZ34oifb8h0vxeB_6yAILgf4a71WFC5mwQc">

I think I would not want to remove these metadata torrents in the slow-check. There's a specific script for that, and I would suggest that we amend that one if it fails to capture metadata items.

In the file "remove_metadata_missing.py", you see that this condition is there to capture items that fail to grab metadata:
`if  queueItem['status'] == 'queued' and queueItem['errorMessage'] == 'qBittorrent is downloading metadata':`. However, that status is "queued", not downloading.

I'd like to understand why your queue-json does not say anything about metadata. Could you post the properties of the torrent that you see in qbit? Also, for how long are they in "downloading" state as opposed to "queued"?

Also, to make sure these metadata torrents (in downloading state) are not coming up in the slow check, I think what we can easily do is to change the below to `if queueItem['status'] == 'downloading' and queueItem['size'] `: 

https://github.com/ManiMatter/decluttarr/blob/2ee86d35914d24318b83c0672add7954571c9e0e/src/jobs/remove_slow.py#L30

slkiny commented:
Thanks for the feedback, yeah this seems like an edge case either the metadata missing or slow could detect.

I don't have a personal preference, but seems that you'd prefer it to be in the metadata handler.

I saw the metadata code was more about queued or errors. While this metadata seems to be the system thinks it can download it but cannot finish downloading it so I made the change in the slow download section.

Regardless of the final placement, another phrasing of this issue could be 

**"To Detect (slow) downloads that are invalid and have a size of 0 bytes for over X minutes, and remove those"**

With that insight in mind is why I picked to put it in slow download, but i can see about moving it if you feel strongly about it.

slkiny commented:
If you agree with the above statement we can modify the statement of the logs to be more about, a download with 0 size was detected. This looks malformed and if it doesn't progress or get a size > 0 soon we will terminate it etc..

**"A slow download with 0 size was detected. This looks malformed and if it doesn't progress or get a size > 0 soon, we will terminate it."**

slkiny commented:
By the way thanks for this project, I made my own prior. And found this to be a lot better :)

ManiMatter commented:
Hi

thanks for your thoughts.

I am still unsure that I understand why this scenario even occurs. On the qbit screenshot that you posted  the status is ‚Downloading metadata‘, but the json that you posted does not carry the status.

For me, slow check is for stuff in progress; metadata is a different gig, hence why I would put it over there. But rather to piecemeal together that the downloading of metadata is taking forever, it would be great if that status came through in the queue response. Maybe the sonarr team could add it and then we add that specific status to the metadata check?

ManiMatter commented:
I still think we need to get to the bottom why your queue response is ‚downloading‘ (instead of ‚queued‘) and says nothing about metadata. 

> On the json that you posted, they are in `'status': 'downloading', 'trackedDownloadStatus': 'ok', 'trackedDownloadState': 'downloading', 'statusMessages': []`. So no mentioning of metdata.

After all, there are other downloads that are marked like that when downloading metadata. 

https://github.com/ManiMatter/decluttarr/blob/2ee86d35914d24318b83c0672add7954571c9e0e/src/jobs/remove_metadata_missing.py#L17

Maybe worthwhile asking the sonarr team?

slkiny commented:
Yeah I tried looking into it more, but its still a curious bug; maybe related to how i tend to Force Download things possibly? I moved it to the metadata piece as suggested, but i dont see any harm in having it as a double check in general.

Also i added a shared func to check if the keys exist in a dict. A bit more readable.

ManiMatter commented:
I think you should ask sonarr why the metadata state is not in errorMessage, and why the state is downloading, not queued. Because that is how it normally is; if they then might realize that they have not added the metadata state for downloading state, they add it and we use that.

ManiMatter commented:
@slkiny - are you still looking into this?

ManiMatter commented:
In my view its not a problem of ‚malformed torrents‘ but that the radarr/sonarr app for some reason do not pass on the torrent status.

have you considered this suggestion?


> I think you should ask sonarr why the metadata state is not in errorMessage, and why the state is downloading, not queued. Because that is how it normally is; if they then might realize that they have not added the metadata state for downloading state, they add it and we use that.
> 
> Refer to this: [Sonarr/Sonarr#5940](https://github.com/Sonarr/Sonarr/issues/5940)
> 
> Essentially what they achieved here is to differentiate between truly queued downloads, and those fetching metadata, by adding `"statusMessages": ["qBittorrent is downloading metadata"]`.
> 
> It seems that you have hit a scenario where the status is not `queued` but `downloading` and the same differentiation is required. Maybe they can build on what they have already done for `queued`?

ManiMatter commented:
ping?

slkiny commented:
Hey, in general I think defense in depth is key. So if the upstream is stuck regardless of reason or status message that we should stop it/ remove it. 

I'm happy with the current state and or the prior when we checked download status amount. Even if it gets fixed upstream it's always good to have double checks since then if there is a regression etc the program handles it regardless

ManiMatter commented:
I see it differently; I would always suggest to first fix issues at source before implementing a workaround.

Particularly true with sonarr/radarr because they usually are quite responsive and helpful.

Please raise it with them as suggested above. Only if they don‘t move, I will consider merging this PR

ManiMatter commented:
> In my view its not a problem of ‚malformed torrents‘ but that the radarr/sonarr app for some reason do not pass on the torrent status.
> 
> have you considered this suggestion?
> 
> > I think you should ask sonarr why the metadata state is not in errorMessage, and why the state is downloading, not queued. Because that is how it normally is; if they then might realize that they have not added the metadata state for downloading state, they add it and we use that.
> > Refer to this: [Sonarr/Sonarr#5940](https://github.com/Sonarr/Sonarr/issues/5940)
> > Essentially what they achieved here is to differentiate between truly queued downloads, and those fetching metadata, by adding `"statusMessages": ["qBittorrent is downloading metadata"]`.
> > It seems that you have hit a scenario where the status is not `queued` but `downloading` and the same differentiation is required. Maybe they can build on what they have already done for `queued`?

Btw, just as another pointer. In Sonarr's code I see they do a specific handling for metadata when not forced:
https://github.com/Sonarr/Sonarr/blob/e97e5bfe8f3a7f01f7188f786eb7b51f1d8ef8b5/src/NzbDrone.Core/Download/Clients/QBittorrent/QBittorrent.cs#L283-L293

But I don't see any specific handling when it's forced. I'm not sure if this is of relevance for your problem, but maybe an extra input if you plan to you raise it with them.
https://github.com/Sonarr/Sonarr/blob/e97e5bfe8f3a7f01f7188f786eb7b51f1d8ef8b5/src/NzbDrone.Core/Download/Clients/QBittorrent/QBittorrent.cs#L298

If you don't plan to raise it with them, please close this issue.

ManiMatter commented:
https://github.com/Sonarr/Sonarr/pull/7001/commits

ManiMatter commented:
Considered fixed at source. Closing.
